test/retry_test.go

```diff
@@ -102,6 +102,64 @@ func (s) TestRetryUnary(t *testing.T) {
 	}
 }
 
+func (s) TestRetryMaxAttemptsUnary(t *testing.T) {
+	callCount := 0
+	ss := &stubserver.StubServer{
+		EmptyCallF: func(context.Context, *testpb.Empty) (r *testpb.Empty, err error) {
+			defer func() { t.Logf("server call %v returning err %v", callCount, err) }()
+			callCount++
+			return nil, status.New(codes.AlreadyExists, "retryable error").Err()
+		},
+	}
+	testCases := []struct {
+		serviceMaxAttempts int
+		clientMaxAttempts  int
+		expectedAttempts   int
+	}{
+		{serviceMaxAttempts: 9, clientMaxAttempts: 4, expectedAttempts: 4},
+		{serviceMaxAttempts: 9, clientMaxAttempts: 7, expectedAttempts: 7},
+		{serviceMaxAttempts: 3, clientMaxAttempts: 10, expectedAttempts: 3},
+		{serviceMaxAttempts: 8, clientMaxAttempts: -1, expectedAttempts: 5}, // 5 is default max
+		{serviceMaxAttempts: 3, clientMaxAttempts: 0, expectedAttempts: 1},
```

Comment:
How? :)

Reply:
I agree this visually doesn't make sense. IIRC, retry logic isn't hit until after the first attempt so you'll always get at least one attempt. I did not change any of the retry logic, I just added the test that exposed this behavior.

This test will change once I implement your suggested lower bound on the setting. The dial option name probably makes more sense as `maxAttempts` rather than `maxRetryAttempts`. I will change the name.

Reply:
I was mostly joking here and referencing my other comment.  :)

The only change I think the test will need is to remove the `if` around the place where you create the DialOption.

It's true that "`retry`" doesn't really explain this fully, but `MaxAttempts` seems too nebulous.  How about `WithMaxCallAttempts`?

Reply:
sgtm